src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -71,63 +71,200 @@ public static unsafe int SequenceCompareTo(ref char first, int firstLength, ref
             }
 
 #if BIT64
-            if ((byte*)minLength >= (byte*)(i + sizeof(int) / sizeof(char)))
+            ulong difference;
+            while (minLength >= (offset + sizeof(ulong) / sizeof(char)))
             {
-                if (Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) ==
-                    Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
+                ulong firstLong = Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, (IntPtr)offset)));
+                ulong secondLong = Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, (IntPtr)offset)));
+
+                difference = firstLong ^ secondLong;
+                if (difference != 0)
                 {
-                    i += sizeof(int) / sizeof(char);
+                    goto LongDifference;
                 }
+
+                offset += sizeof(ulong) / sizeof(char);
             }
 #endif
+            if (minLength >= (offset + sizeof(int) / sizeof(char)))
+            {
+                if (Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, (IntPtr)offset))) ==
+                    Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, (IntPtr)offset))))
+                {
+                    offset += sizeof(int) / sizeof(char);
+                }
+            }
 
-            while ((byte*)i < (byte*)minLength)
+            int result = 0;
+            while (offset < minLength)
             {
-                int result = Unsafe.Add(ref first, i).CompareTo(Unsafe.Add(ref second, i));
+                result = Unsafe.Add(ref first, (IntPtr)offset).CompareTo(Unsafe.Add(ref second, (IntPtr)offset));
                 if (result != 0)
-                    return result;
-                i += 1;
+                    goto ResultDifference;
+                offset++;
             }
 
         Equal:
-            return lengthDelta;
+            // At this point, we have compared all the characters in at least one string.
+            // The longer string will be larger; if they are the same length this will return 0.
+            return firstLength - secondLength;
+
+#if BIT64
+        LongDifference:
+            // Find bit offset of first difference and add to current offset,
+            // convert the bit position to char position by dividing by (2 * 8) for chars (shift right by 4)
+            offset += BitOperations.TrailingZeroCount(difference) >> 4;
+#endif
+        OffsetDifference:
+            result = Unsafe.Add(ref first, (IntPtr)offset).CompareTo(Unsafe.Add(ref second, (IntPtr)offset));
+            Debug.Assert(result != 0);
+
+        ResultDifference:
+            return result;
+
+        IntrinsicsCompare:
+            if (Sse2.IsSupported)
+            {
+                uint matches;
+                nint lengthToExamine;
+                if (Avx2.IsSupported)
```

Comment:
Is an AVX2 special-case worth it over using SSE everywhere? We didn't AVX2-enable the ASCII / UTF-8 handling code paths because they didn't show a significant enough benefit to warrant the extra work.
/cc @tannergooding

Reply:
AVX2 for widen is a different as you need to add extra instructions to also permute?

Reply:
I think @GrabYourPitchforks is asking what is the perf difference between the 128-bit (with VEX encoding) and 256-bit codepaths?

Reply:
AVX2 is 1 : 2 SSE instructions here; whereas for widen its 2 : 2 instructions (due to the need for permute in the AVX2 path); so the number of instructions that need to be run is not reduced so can see it not providing any improvement there; here it halves the number of instructions run.

Reply:
e.g. the main AVX2 loop is
```asm
G_M47651_IG15:
       lea      rax, [r10+r10]
       vmovupd  ymm0, ymmword ptr[rcx+rax]
       vmovupd  ymm1, ymmword ptr[r8+rax]
       vpcmpeqw ymm0, ymm0, ymm1
       vpmovmskb eax, ymm0
       cmp      eax, -1
       jne      G_M47651_IG20
       add      r10, 16
       cmp      r11, r10
       ja       SHORT G_M47651_IG15
```

Reply:
@AndyAyersMS any reason its
```asm
G_M47651_IG15:
       lea      rax, [r10+r10]
       vmovupd  ymm0, ymmword ptr[rcx+rax]
       vmovupd  ymm1, ymmword ptr[r8+rax]
```
rather than 
```asm
G_M47651_IG15:
       vmovupd  ymm0, ymmword ptr[rcx+2*r10]
       vmovupd  ymm1, ymmword ptr[r8+2*r10]
```

Seems to be happy to do it in other places
```asm
G_M47651_IG09:
       lea      r11, bword ptr [rcx+2*r10]    ; 2*
       movzx    rsi, word  ptr [r8+2*r10]    ; 2*
       movzx    r11, word  ptr [r11]
       sub      r11d, esi
       jne      SHORT G_M47651_IG12
       inc      r10
       cmp      r10, rax
       jb       SHORT G_M47651_IG09

G_M47651_IG10:
       mov      r11d, edx
       sub      r11d, r9d
       jmp      SHORT G_M47651_IG12

G_M47651_IG11:
       lea      rax, bword ptr [rcx+2*r10]    ; 2*
       movzx    rdx, word  ptr [r8+2*r10]    ; 2*
       movzx    r11, word  ptr [rax]
       sub      r11d, edx
```